### PR TITLE
b/189965239: Allow using another service account for GCS fetches.

### PR DESCRIPTION
**Pull request by toddbeckman-g**
- Allow choosing between default credentials or a provided service account.
- Use IAM Credentials API to impersonate the provided service account.

Some additional refactoring:
- Streamline client interfaces for easier testing.
- Cover some testing edge cases which were missed.

### CODE REVIEW

File: src/go/gcsrunner/serviceaccount.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+	"cloud.google.com/go/storage"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+)
+
+// staticTokenSource implements oath2.TokenSource
+type staticTokenSource struct {
+	accessToken string    `json:"access_token"`
+	tokenType   string    `json:"token_type,omitempty"`
```
Comment (KyleWiese) [inline src/go/gcsrunner/serviceaccount.go:20]: Do we use `tokenType` anywhere?

File: src/go/gcsrunner/serviceaccount.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+func (s staticTokenSource) Token() (*oauth2.Token, error) {
+	return &oauth2.Token{
+		AccessToken: s.accessToken,
+		Expiry:      s.expiry,
+	}, nil
+}
+
+type iamcredentialsResponse struct {
```
Comment (KyleWiese) [inline src/go/gcsrunner/serviceaccount.go:31]: Optional: we could also use https://pkg.go.dev/google.golang.org/api/iamcredentials/v1#GenerateAccessTokenResponse here instead of rolling our own response struct.

File: src/go/gcsrunner/serviceaccount.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	if err != nil {
+		return nil, err
+	}
+	token, err := c.TokenSource.Token()
+	if err != nil {
+		return nil, err
+	}
+	token.SetAuthHeader(req)
+
+	resp, err := http.DefaultClient.Do(req)
```
Comment (KyleWiese) [inline src/go/gcsrunner/serviceaccount.go:63]: Should we specify an HTTP client with a timeout? e.g 5 or 10 secs

File: src/go/gcsrunner/serviceaccount.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed iamcredentials request: %v", resp.Status)
```
Comment (KyleWiese) [inline src/go/gcsrunner/serviceaccount.go:68]: Might be helpful to also print the error body here (if present)

File: src/go/gcsrunner/main/runner.go [hunk 5]
```diff
@@ -52,6 +55,7 @@ var (
 	envoyLogPath = flag.String("envoy_log_path", "",
 		"Envoy application logging path. Default is to write to stderr.")
 	envoyComponentLogLevel = flag.String("envoy_component_log_level", "", "Mapping for Envoy log level by component.")
+	sa                     = flag.String("run_as_service_account", "", "If provided, use this account when fetching the config from GCS.")
```
Comment (KyleWiese) [inline src/go/gcsrunner/main/runner.go:58]: nit: "If provided, use this account when fetching the config from GCS. If not specified, the instance's default credentials will be used. "

File: src/go/gcsrunner/fetch_config_test.go [hunk 6]
```diff
@@ -15,147 +15,193 @@
...
 		},
 		{
-			name:                    "success with retries",
-			wantBytes:               output,
-			wantMultipleReaderCalls: true,
-			passMultipleReaderCalls: true,
-			newReaderReturns:        bytes.NewReader(output),
-			newReaderErr:            fmt.Errorf("temporary reader error should resolve"),
-			credsFunc:               goodFindDefaultCredentials,
+			name:           "timeout retrying newGSC",
```
Comment (KyleWiese) [inline src/go/gcsrunner/fetch_config_test.go:160]: s/newGSC/newGCS

File: src/go/gcsrunner/fetch_config.go [hunk 7]
```diff
@@ -27,52 +27,67 @@ import (
...
+	ConfigFileName                string
+	WriteFilePath                 string
+	FetchGCSObjectInitialInterval time.Duration
+	FetchGCSObjectTimeout         time.Duration
 }
 
 var (
-	osCreate = func(p string) (file, error) { return os.Create(p) }
+	// can be overwritten in unit tests
+	newGCSClient = func(ctx context.Context, sa string) (gcsReader, error) {
```
Comment (KyleWiese) [inline src/go/gcsrunner/fetch_config.go:69]: optional: Instead of making these function vars, they could just be functions e.g

```
var (
 gcsCient = newGCSClient
)
func newGCSClient(...)...
```

File: src/go/gcsrunner/serviceaccount.go (hunk 2 above)
Comment (toddbeckman-g) [reply src/go/gcsrunner/serviceaccount.go:30]: The struct here versus in the iamcredentials API is functionally identical anyway, so I don't think we gain much by including this dependency. The context here is that my attempts to use the iamcredentials API resulted in odd errors I couldn't debug, and when I stepped away from using the API at all, everything just worked™.

**Review by KyleWiese — APPROVED** [review]

**Review by TAOXUY — APPROVED** [review]
